src/tagsQLController.js

```diff
@@ -0,0 +1,395 @@
+// "Natural language" operators
+let OPERATOR_EQ = '=';
+let OPERATOR_NOTEQ = '!=';
+let OPERATOR_IN = 'is in';
+let OPERATOR_NOTIN = 'is not in';
+let OPERATOR_EXISTS = 'exists';
+let OPERATOR_NOTEXISTS = 'doesn\'t exist';
+let OPERATOR_AND = 'AND';
+let OPERATOR_OR = 'OR';
+
+export class TagsQLController {
+
+  constructor(uiSegmentSrv, datasource, $q, targetSupplier)  {
+    this.uiSegmentSrv = uiSegmentSrv;
+    this.datasource = datasource;
+    this.$q = $q;
+    this.targetSupplier = targetSupplier;
+    this.removeTagSegment = uiSegmentSrv.newSegment({fake: true, value: '-- Remove tag --'});
+    this.removeValueSegment = uiSegmentSrv.newSegment({fake: true, value: '-- Remove value --'});
+  }
+
+  initTagsSegments() {
+    let target = this.targetSupplier();
+    if (!target.tagsQL && target.tags) {
+      // Compatibility, switching from older version
+      target.tagsQL = convertFromKVPairs(target.tags);
+    }
+    var segments = stringToSegments(target.tagsQL, this.uiSegmentSrv);
+    segments.push(this.uiSegmentSrv.newPlusButton());
+    // Fix plus-button: add it at the end of each enumeration
+    var isInEnum = false;
+    for (var i = 0; i < segments.length; i++) {
```

Comment:
I was not sure since I modify the array during the loop (splice).... That's why I preferred to keep an explicit increment index